src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -19,6 +19,23 @@ function rrule(::typeof(dot), x, y)
     return dot(x, y), dot_pullback
 end
 
+function frule((_, Δx, ΔA, Δy), ::typeof(dot), x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+    return dot(x, A, y), dot(Δx, A, y) + dot(x, ΔA, y) + dot(x, A, Δy)
+end
+
+function rrule(::typeof(dot), x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
```

Comment:
Could you please restrict these to maybe `StridedArray`s? See https://github.com/JuliaDiff/ChainRules.jl/issues/232 for context.

Reply:
And by `StridedArray`s I actually mean `StridedArray{<:Real}`.

Reply:
Oh man, there are a lot of words there, maybe later. But this correct and tested for complex numbers. Are you concerned about `dot`'s methods for Diagonal etc?

Reply:
Haha fair enough. The code you've written and tested certainly looks like it's correct to me, so I'm happy with that.

Yeah, the primary concern is about things like `Diagonal`, or really any subtypes of `AbstractMatrix` / `AbstractVector` for which this `rrule` would yield the wrong asymptotic complexity.

I agree regarding restrictiveness -- the best solution that we have for now is to to explicitly enable particular types. So, for example, it would be fine to explicitly allow for eg. `transpose`d arrays.

It's all very unsatisfactory :(

Reply:
Agree that projecting onto UpperTriangular etc. is probably the right thing to do, but ideally done uniformly at a higher level, not written into every single rule? Or by adding some function `project_like(dA, A)` to every rule, starting with `*`, presumably (there were some Zygote issues about this). Likewise projecting onto the real line, for real parameters (ditto). For some there may be efficiency gains to specialising the rule, but often N^2 dense will beat 1/2 N^2 sparse.

I've added a method for Diagonal, for which there's an obvious complexity gain. I suppose something like `dot(::Fill, ::Fill)` would also change complexity, but when someone feels the need for that, maybe that's the time to add it to both 2- and 3-arg cases.

I still think restricting to StridedMatrix is much too narrow, e.g. surely this ought to work with SMatrix input? As Zygote currently does. And as the rule for `dot(::Any, ::Any)` allows too.

Reply:
> I still think restricting to StridedMatrix is much too narrow, e.g. surely this ought to work with SMatrix input? As Zygote currently does. And as the rule for dot(::Any, ::Any) allows too.

In the `SMatrix` case, Zygote ought to be able to handle that without any rules since there isn't any mutation.

So I just tried out dot with both a `Diagonal{Float64}` and `Matrix{Float64}` using your implementation but restricting the types to strided vectors / matrix, and it seems to work fine. In this case Zygote was able to backprop through the rules nicely. Does that put your doubts to rest?

Incidentally that means that we don't need the `Diagonal` rule, which is good news.

Reply:
Tagged version:
```julia
julia> using Zygote, StaticArrays, BenchmarkTools, Random

julia> Random.seed!(0); sm = @SMatrix rand(10,10); sv = @SVector rand(10);

julia> gradient(m -> real(dot(sv,m,sv)), sm)[1]
ERROR: Can't differentiate loopinfo expression
Stacktrace:
 [1] error(::String) at /Applications/Julia-1.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 [2] macro expansion at ./simdloop.jl:79 [inlined]
 [3] dot at /Applications/Julia-1.5.app/Contents/Resources/julia/share/julia/stdlib/v1.5/LinearAlgebra/src/generic.jl:936 [inlined]
 [4] (::typeof(∂(dot)))(::Float64) at /Users/me/.julia/packages/Zygote/seGHk/src/compiler/interface2.jl:0
...

julia> @btime gradient(m -> real(dot($sv,m,$sv)), Diagonal($sv))[1]
  1.504 ms (5532 allocations: 242.83 KiB)
(diag = [0.5106375128426506, 0.19312327745527802, 0.07834359296248952, 0.11572530436676702, 0.05607417133446614, 0.005420958927950359, 0.012818381211631749, 0.7835180639610391, 0.26488415942189886, 0.48020847299642594],)

julia> typeof(ans)                                                            
NamedTuple{(:diag,),Tuple{MArray{Tuple{10},Float64,1,10}}}

julia> v = rand(100);

julia> @btime gradient(x -> dot($v, x, $v), Diagonal($v))[1] isa Diagonal
  10.512 ms (46239 allocations: 2.19 MiB)
false
```

With this PR:
```julia
julia> @btime gradient(m -> real(dot($sv,m,$sv)), $sm)[1] isa SMatrix
  89.729 ns (0 allocations: 0 bytes)
true

julia> @btime gradient(m -> real(dot($sv,m,$sv)), Diagonal($sv))[1]
  28.701 ns (0 allocations: 0 bytes)
10×10 Diagonal{Float64,SArray{Tuple{10},Float64,1,10}}:
 0.510638   ⋅         ⋅          ⋅        …   ⋅         ⋅         ⋅ 
  ⋅        0.193123   ⋅          ⋅            ⋅         ⋅         ⋅ 
  ⋅         ⋅        0.0783436   ⋅            ⋅         ⋅         ⋅ 
  ⋅         ⋅         ⋅         0.115725      ⋅         ⋅         ⋅ 
  ⋅         ⋅         ⋅          ⋅            ⋅         ⋅         ⋅ 
  ⋅         ⋅         ⋅          ⋅        …   ⋅         ⋅         ⋅ 
  ⋅         ⋅         ⋅          ⋅            ⋅         ⋅         ⋅ 
  ⋅         ⋅         ⋅          ⋅           0.783518   ⋅         ⋅ 
  ⋅         ⋅         ⋅          ⋅            ⋅        0.264884   ⋅ 
  ⋅         ⋅         ⋅          ⋅            ⋅         ⋅        0.480208

julia> log2(1_504_000 / 28.7) * 1.5  # years
23.5160914568271

julia> @btime gradient(x -> dot($v, x, $v), Diagonal($v))[1] isa Diagonal    
  545.546 ns (16 allocations: 3.19 KiB)
true

julia> log2(10_512_000 / 545.546) * 1.5
21.350965209969427
```

Behaviour of `*`, out of scope for this PR:
```julia
julia> gradient(m -> sum(exp, sm * m), Diagonal(sv))[1] |> typeof            
SArray{Tuple{10,10},Float64,2,100}

julia> Diagonal(sv) * Diagonal(sv) |> typeof
Diagonal{Float64,SArray{Tuple{10},Float64,1,10}}

julia> gradient(m -> sum(exp, Diagonal(sv) * m), Diagonal(sv))[1] |> typeof
Array{Float64,2}
```

Reply:
Interesting that this doesn't work with StaticArrays. I would be interested to understand why.

As regards the `Diagonal` -- I guess the performance improvement is unsurprising because Zygote's type inference tends to fall over.

The type of the gradient obtained for the `Diagonal` with the naive Zygote implementation is fine -- it's a `NamedTuple`, which is an appropriate type for the gradient of a `Diagonal` matrix.

Anyway, that being said, I'm happy to accept the `AbstractMatrix` / `AbstractVector` proposal for now, since we don't have a uniformly better alternative.

Would you be willing to restrict the element type to `Number`?

Reply:
I'm not always sure why, but asking Zygote to do as little as possible is great for performance. `dot(x,D::Diagonal,y)` is `mapreduce(t -> f(t[1], t[2], t[3]), +, zip(x, D.diag, y))`, I guess it's possible to imagine a compiler sufficiently smart to make the backward code efficient, but presumably it's always going to be much harder to guess what it's tripping over today, than it is in simple code you wrote. (It's also harder to imagine that working with Tracker/ReverseDiff though, and IIRC the plan is eventually to also use rules for such systems.) StaticArrays seems even harder, a lot of human effort has gone into helping it be fast on the functions it supports, so either you hit more such functions on the reverse, or you don't benefit. This seems like a major part of the reason to define rules at all, beyond a few basic laws.

Probably there should be tests with StaticArrays, at some point if not this PR?

Agree NamedTuples are fine internally, although confusing if returned for user input.

Can change to Number, sure. I guess this also assumes `*` is commutative.

Reply:
BTW, attempting to add tests like
```julia
        permuteddimsarray(A) = PermutedDimsArray(A, (2,1))
        @testset "$F{$T}" for T in (Float64, ComplexF64), F in (adjoint, permuteddimsarray)
            M, N = 3, 4
            x, A, y = randn(T, M), F(randn(T, N, M)), randn(T, N)
            ẋ, Adot, ẏ = randn(T, M), F(randn(T, N, M)), randn(T, N)
            x̄, Abar, ȳ = similar(x), F(randn(T, N, M)), similar(y)
            frule_test(dot, (x, ẋ), (A, Adot), (y, ẏ))
            rrule_test(dot, randn(T), (x, x̄), (A, Abar), (y, ȳ))
        end
```
gives an error like 
```
  Got exception outside of a @test
  MethodError: Cannot `convert` an object of type Array{Complex{Float64},1} to an object of type Base.ReshapedArray{Complex{Float64},1,PermutedDimsArray{Complex{Float64},2,(2, 1),(2, 1),Array{Complex{Float64},2}},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}}
...
  Stacktrace:
   [1] oftype(::Base.ReshapedArray{Complex{Float64},1,PermutedDimsArray{Complex{Float64},2,(2, 1),(2, 1),Array{Complex{Float64},2}},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}}, ::Array{Complex{Float64},1}) at ./essentials.jl:367
   [2] (::FiniteDifferences.var"#Vector_from_vec#42"{Base.ReshapedArray{Complex{Float64},1,PermutedDimsArray{Complex{Float64},2,(2, 1),(2, 1),Array{Complex{Float64},2}},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}},Array{Array{Float64,1},1},Array{FiniteDifferences.var"#Complex_from_vec#40",1}})(::Array{Float64,1}) at /Users/me/.julia/packages/FiniteDifferences/sCBeL/src/to_vec.jl:30

   [19] frule_test(::Function, ::Tuple{Array{Complex{Float64},1},Array{Complex{Float64},1}}, ::Vararg{Tuple{Any,Any},N} where N) at /Users/me/.julia/packages/ChainRulesTestUtils/9hZgi/src/testers.jl:184
```
so perhaps FiniteDifferences is accepting AbstractArray when it wants Array?

Reply:
Yeah, that's to do with `to_vec`. We currently have to define it for every new type that we want to work with.

[Here's a PR](https://github.com/JuliaDiff/FiniteDifferences.jl/pull/108) that should resolve the issue.

Reply:
OK, I stuck in the above tests, which now work at least locally. See if CI likes it.